File: java/HexDump.java
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

public class HexDump {
    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("usage: HexDump filename");
            return;
        }

        try (InputStream input = getInputStream(args)) {
            int position = 0;
            byte[] buffer = new byte[16];

            while (true) {
                int bytesRead = input.read(buffer, 0, buffer.length);
                if (bytesRead == -1) {
                    return;
                }

                System.out.printf("%04x: ", position);
                position += bytesRead;

                for (int i = 0; i < 16; i++) {
                    if (i < bytesRead) {
                        System.out.printf("%02x ", buffer[i] & 0xFF);
                    } else {
                        System.out.print("   ");
                    }
                    if (i == 7) {
                        System.out.print("-- ");
                    }
                }

                for (int i = 0; i < bytesRead; i++) {
                    int value = buffer[i] & 0xFF;
                    if (value < 32) {
                        value = '?';
                    }
                    System.out.print((char) value);
                }

                System.out.println();
            }
        } catch (IOException ex) {
            System.err.println("I/O error while dumping input: " + ex.getMessage());
        }
    }

    private static InputStream getInputStream(String[] args) {
        if ((args.length != 1) || !new java.io.File(args[0]).exists()) {
            return System.in;
        }

        try {
            return new FileInputStream(args[0]);
        } catch (SecurityException | IOException ex) {
            System.err.printf("Unable to read %s, dumping from stdin: %s%n", args[0], ex.getMessage());
            return System.in;
        }
    }
}

File: csharp/HexDump.cs
if (args.Length != 1)
{
    Console.WriteLine("usage: HexDump filename");
    return;
}

using (var input = GetInputStream(args))
{
    var position = 0;
    var buffer = new byte[16];
    while (true)
    {
        var bytesRead = input.Read(buffer, 0, buffer.Length);
        if (bytesRead == 0) return;

        Console.Write("{0:x4}: ", position);
        position += bytesRead;

        for (var i = 0; i < 16; i++)
        {
            if (i < bytesRead)
                Console.Write("{0:x2} ", (byte)buffer[i]);
            else
                Console.Write("   ");
            if (i == 7) Console.Write("-- ");
        }

        for (int i = 0; i < bytesRead; i++)
        {
            if (buffer[i] < 32)
                buffer[i] = (byte)'?';
            Console.Write((char)buffer[i]);
        }
        Console.WriteLine();
    }
}

Stream GetInputStream(string[] args)
{
    if ((args.Length != 1) || !File.Exists(args[0]))
        return Console.OpenStandardInput();
    else
    {
        try
        {
            return File.OpenRead(args[0]);
        }
        catch (UnauthorizedAccessException ex)
        {
            Console.Error.WriteLine("Unable to read {0}, dumping from stdin: {1}",
                                    args[0], ex.Message);
            return Console.OpenStandardInput();
        }
    }
}
